File: server/src/main/java/com/walkietalkie/protocol/AudioFrame.java
package com.walkietalkie.protocol;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public final class AudioFrame {
    public static final int HEADER_BYTES = 13;
    private AudioFrame() {}

    public static boolean isValid(byte[] frame, int maxBytes) {
        return frame.length >= HEADER_BYTES && frame.length <= maxBytes && (frame[0] & 0xff) == Protocol.VERSION;
    }

    public static long sequence(byte[] frame) { return Integer.toUnsignedLong(ByteBuffer.wrap(frame, 1, 4).order(ByteOrder.BIG_ENDIAN).getInt()); }
    public static long timestampMillis(byte[] frame) { return ByteBuffer.wrap(frame, 5, 8).order(ByteOrder.BIG_ENDIAN).getLong(); }
}

File: app/src/main/java/com/walkietalkie/app/AudioFrame.kt
package com.walkietalkie.app

import java.nio.ByteBuffer
import java.nio.ByteOrder

object AudioFrame {
    private const val VERSION: Byte = 1
    private const val HEADER_BYTES = 13

    fun encode(sequence: Long, timestampMillis: Long, opusPacket: ByteArray): ByteArray {
        require(sequence in 0..0xffffffffL)
        require(opusPacket.isNotEmpty())
        return ByteBuffer.allocate(HEADER_BYTES + opusPacket.size).order(ByteOrder.BIG_ENDIAN)
            .put(VERSION).putInt(sequence.toInt()).putLong(timestampMillis).put(opusPacket).array()
    }
}
